File: 907-Sum_Of_Subarray_Minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        n = len(arr)
        prevSmaller = [-1] * n
        monoStack = deque()
        res = [0] * n

        for i in range(n):
            while monoStack and arr[monoStack[-1]] > arr[i]:
                monoStack.pop()

            if monoStack:
                prevSmaller[i] = monoStack[-1]
            monoStack.append(i)

        for i in range(n):
            j = prevSmaller[i]
            res[i] = res[j] + arr[i]*(i-j)

        return sum(res) % (10**9 + 7)
        

File: 907-Sum_Of_Subarray_Minimums.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var sumSubarrayMins = function(arr) {
    let n = arr.length;
    let prevSmaller = new Array(n).fill(-1);
    let res = new Array(n).fill(0);
    let monoStack = [];
    
    for (let i=0; i < n; i++){
        while (monoStack.length && arr[monoStack.at(-1)] > arr[i]) monoStack.pop();
        
        if (monoStack.length) prevSmaller[i] = monoStack.at(-1);
        monoStack.push(i);
    }

    for (let i=0; i < n; i++) res[i] = res.at(prevSmaller[i]) + arr[i]*(i-prevSmaller[i]);
    return res.reduce((a, b) => {return a+b}, 0) % (10**9+7);
};

